tests/drivers/i2c/i2c_slave_eeprom/src/main.c

```diff
@@ -0,0 +1,212 @@
+		to_display_format(eeprom_data, TEST_DATA_SIZE,
+				  buffer_print_eeprom);
+		SYS_LOG_ERR("Buffer contents are different: %s",
+			    buffer_print_i2c);
+		SYS_LOG_ERR("                           vs: %s",
+			    buffer_print_eeprom);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int run_partial_read(struct device *i2c, unsigned int offset)
+{
+	int ret;
+
+	SYS_LOG_INF("Start partial read from EEPROM (off=%d)", offset);
+
+	ret = i2c_burst_read(i2c, CONFIG_I2C_SLAVE_EEPROM_ADDRESS,
+			     offset, i2c_buffer, TEST_DATA_SIZE-offset);
+	if (ret) {
+		SYS_LOG_ERR("Failed to read EEPROM Code %d", ret);
+		return -1;
+	}
+
+	if (memcmp(i2c_buffer, &eeprom_data[offset], TEST_DATA_SIZE-offset)) {
+		to_display_format(i2c_buffer, TEST_DATA_SIZE-offset,
+				  buffer_print_i2c);
+		to_display_format(&eeprom_data[offset], TEST_DATA_SIZE-offset,
+				  buffer_print_eeprom);
+		SYS_LOG_ERR("Buffer contents are different: %s",
+			    buffer_print_i2c);
+		SYS_LOG_ERR("                           vs: %s",
+			    buffer_print_eeprom);
+		return -1;
+	}
+
+	return 0;
+}
+
+static int run_program_read(struct device *i2c, unsigned int offset)
+{
+	int ret, i;
+
+	SYS_LOG_INF("Start program of EEPROM (off=%d)", offset);
+
+	for (i = 0 ; i < TEST_DATA_SIZE-offset ; ++i)
+		i2c_buffer[i] = i;
+
+	ret = i2c_burst_write(i2c, CONFIG_I2C_SLAVE_EEPROM_ADDRESS,
+			      offset, i2c_buffer, TEST_DATA_SIZE-offset);
+
+	if (ret) {
+		SYS_LOG_ERR("Failed to write EEPROM Code %d", ret);
+		return -1;
+	}
+
+	memset(i2c_buffer, 0xFF, TEST_DATA_SIZE);
+
+	/* Read back EEPROM from I2C Master requests, then compare */
+	ret = i2c_burst_read(i2c, CONFIG_I2C_SLAVE_EEPROM_ADDRESS,
+			     offset, i2c_buffer, TEST_DATA_SIZE-offset);
+	if (ret) {
+		SYS_LOG_ERR("Failed to read EEPROM Code %d", ret);
+		return -1;
+	}
+
+	for (i = 0 ; i < TEST_DATA_SIZE-offset ; ++i) {
+		if (i2c_buffer[i] != i) {
+			to_display_format(i2c_buffer, TEST_DATA_SIZE-offset,
+					  buffer_print_i2c);
+			SYS_LOG_ERR("Invalid Buffer content: %s",
+				    buffer_print_i2c);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+void main(void)
+{
+	struct device *eeprom;
+	struct device *i2c;
+	const struct eeprom_slave_api *funcs;
+	int ret, offset;
+
+	i2c = device_get_binding(I2C_MASTER_DEV_NAME);
+	if (i2c) {
+		SYS_LOG_INF("Found I2C device %s", I2C_MASTER_DEV_NAME);
+	} else {
+		SYS_LOG_ERR("I2C device %s not found",
+			    I2C_MASTER_DEV_NAME);
+		return;
+	}
+
+	eeprom = device_get_binding(CONFIG_I2C_SLAVE_EEPROM_NAME);
+	if (eeprom) {
+		SYS_LOG_INF("Found EEPROM device %s", CONFIG_I2C_SLAVE_EEPROM_NAME);
+	} else {
+		SYS_LOG_ERR("EEPROM device %s not found",
+			    CONFIG_I2C_SLAVE_EEPROM_NAME);
+		return;
+	}
+
+	/* Get EEPROM funcs to read/write the internal memory */
+	funcs = i2c_slave_driver_get_funcs(eeprom);
+	if (!funcs) {
+		SYS_LOG_ERR("Failed to get EEPROM funcs");
+		return;
+	}
+
+	/* Program dummy bytes */
+	ret = funcs->program(eeprom, eeprom_data, TEST_DATA_SIZE);
```

Comment:
There should be a way to do this differently.

By exposing a way in the eeprom driver to initialize it with a memory address and the size it's pointing. (so the driver would not own the memory buffer by itself, but an address, etc...). You could have such function in include/drivers/i2c/slave/eeprom.h for instance.

Reply:
Hmm I think I see what you mean, a eeprom_slave function only defined in the eeprom.h header. Indeed, it will simplify a lot.